server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1160,7 +1160,19 @@ public void flush(boolean force, boolean waitIfOngoing) throws EngineException {
      *                      request is detected, no flush will have occurred and the listener will be completed with a marker
      *                      indicating no flush and unknown generation.
      */
-    public abstract void flush(boolean force, boolean waitIfOngoing, ActionListener<FlushResult> listener) throws EngineException;
+    public final void flush(boolean force, boolean waitIfOngoing, ActionListener<FlushResult> listener) throws EngineException {
+        try (var ignored = readLock.acquire()) {
+            ensureOpen();
```

Comment:
I wonder why this additional `ensureOpen` was added? can't we let the `flushHoldingLock` implementation check the open state? Or maybe that was for ReadOnlyEngine to be more correct by returning an ACE?

Reply:
Yes right this is changing the `ReadOnlyEngine` behaviour to check the state properly, making it so that every lock acquire includes an `ensureOpen()` check. That seems a good move to me in its own right, but in fact this is so that we can more easily replace these locks with a nonblocking (refcounting) implementation in https://github.com/elastic/elasticsearch/pull/103123 without introducing that behavioural change.

Reply:
Ok, thanks for the precision!